affinity (#203)
percona/everest-doc: docs/use/affinity.md

#  Leveraging Affinity in Percona Everest

> !!! warning

Reviewer: Maybe it should just be an info or note instead of a warning

Author: @recharte Let's say it is Important.

If you're looking to dive deep into pod anti-affinity, check out the Kubernetes [documentation](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity).

### Set up affinity rules for your cluster

You can set affinity rules in Percona Everest via the **Advanced Configurations** page within the **Affinity** section. 
{.power-number}

1. From the Percona Everest home page, navigate through the **Create Database** wizard until you reach the **Advanced Configurations** page.

2. From the **Advanced Configurations** page, click **Create affinity rule**. The **Add rule** screen will be displayed.

    ![!image](../images/create_affinity.png)

    
3. Enter the information on the **Add Rule** screen. Click **Add rule**. You can view the affinity rule on the components page of the corresponding database.

    ![!image](../images/add_rule_affinity.png)

The following table provides details about the different attributes involved in setting up affinity rules for your databases.

|**Attributes**    |**Description**  |**Note**  |
|----------------|-----------------|--------------|

> |**Components**|<ul><li>Config Server</li><li>DB Node</li><li>Router</br></br></li><li>DB Node</br><br></li><li>DB Node</li><li>Proxy</li></ul>|Applicable for **MongoDB sharded cluster**.<br/></br></br></br></br>Applicable for **MongoDB non sharded cluster**.</br></br>Applicable for **MySQL** and **PostgreSQL** databases.|

Reviewer: I don't feel strongly opinionated about this but I think we could make the table simpler to understand if we go with something like:
* DB node
* Proxy/Router/PGBouncer - Applicable for MySQL, MongoDB sharded clusters and PostgreSQL databases.
* Config Server - Applicable for MongoDB sharded clusters.

|**Priority**|Each type of Affinity can implement two distinct levels of rule enforcement:</br><ul><li>Preffered</br></br></br></br></br></li><li>Required</li></ul>|</br></br>Defines the preferences for Pod scheduling instead of strict requirements. Kubernetes will try to place the Pod according to these preferences, but if they cannot be fulfilled, the Pod will still be scheduled.</br></br>A strict requirement that must be met for a Pod to be scheduled. If the conditions in this field are not met, the Pod will remain unscheduled.|
|**Weight (1-100)**|It prioritizes preferred scheduling rules using a numerical value that indicates how strongly Kubernetes should favor specific placement preferences when scheduling Pods. Higher weights signify stronger preferences, meaning Kubernetes will prioritize fulfilling rules with higher weights over those with lower weights.|Weight is only used when the priority is **Preferred**.|

> |**Topology key**|TopologyKey is label key on nodes that defines the grouping or topology scope (e.g., zone, hostname) where the rules are enforced.| Node affinity rules do not include a topology key. </br></br>Here are some examples of **topologyKey**:</br><ul><li>kubernetes.io/hostname</li><li>topology.kubernetes.io/zone	</li><li>topology.kubernetes.io/region</li><li>custom node labels (e.g., rack)</li></ul>|

Reviewer: ```suggestion
|**Topology key**|TopologyKey is the field that defines the domain for pod placement. The TopologyKey domain (e.g., zone, hostname) is used to determine the placement of the pods being scheduled relative to other pods that match the label rules.| Node affinity rules do not include a topology key. </br></br>Here are some examples of **topologyKey**:</br><ul><li>kubernetes.io/hostname</li><li>topology.kubernetes.io/zone	</li><li>topology.kubernetes.io/region</li><li>custom node labels (e.g., rack)</li></ul>|
```

> |**key**|A key in pod affinity is the label assigned to pods, used to define scheduling rules. This label helps the Kubernetes scheduler identify pods and place new pods in relation to them based on affinity or anti-affinity rules.|The key that you set will be a label present in other existing pods within the cluster, affecting whether your pods will be scheduled on the same nodes as those pods.</br></br>Here are some examples of **key**:</br><ul><li>app</li><li>security</li><li>environment</li><li>custom labels (e.g., security, web-store)</li></ul>|

Reviewer: ```suggestion
|**Key**|A key in pod affinity is the label assigned to pods, used to define scheduling rules. This label helps the Kubernetes scheduler identify pods and place new pods in relation to them based on affinity or anti-affinity rules.|The key that you set will be a label present in other existing pods within the cluster, affecting whether your pods will be scheduled on the same nodes as those pods.</br></br>Here are some examples of **key**:</br><ul><li>app</li><li>security</li><li>environment</li><li>custom labels (e.g., security, web-store)</li></ul>|
```

|**Values**|The values are the specific label values that must match the key for the affinity rule to apply.|Here are some examples for **values**:</br><ul><li>s2</li><li>database</li><li>production</li><li>custom values</li></ul>|

## Use cases for affinity

Here are several detailed use cases for affinity that highlight its diverse applications and the numerous benefits it offers across various contexts.

=== "Node affinity"
    ### Regional scheduling

    Let's consider a use case in which workloads are distributed based on performance requirements, fault tolerance, and regional specifications across designated zones or areas.

>     You need to run a workload in the `us-west2 region` for latency optimaztion and to meet specific compliance requirements.

Reviewer: ```suggestion
    You need to run a workload in the `us-west2` region for latency optimization and to meet specific compliance requirements.
```

    ```sh
    affinity:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
            - key: topology.kubernetes.io/region
            operator: In
            values:
            - "us-west2"

>     ```

Reviewer: Shouldn't we show a screenshot configuring this instead?

Author: I didnt have the build so instead added this...

    ??? info "What happens under the hood"
        - It ensures that the pod is scheduled only on nodes located in the us-west-2 region, as defined by the `topology.kubernetes.io/region` node label.

        - For the rule to be applicable, it is essential that the node possesses a label such as `topology.kubernetes.io/region: us-west-2`. If no nodes correspond to the specified label, the pod will remain in an unscheduled state.

> === "Pods affinity"
>     ### Pods scheduled together
> 
>     Let's consider a use case in which you want to to ensure that HAProxy pods should be scheduled to run on the same Kubernetes node.
>     
>         affinity:
>           podAffinity:
>             requiredDuringSchedulingIgnoredDuringExecution:
>             - labelSelector:
>             matchExpressions:
>                 - key: app
>                 operator: In
>                 values:
>                 - haProxy
>         topologyKey: "kubernetes.io/hostname"
> 
>     ??? info "What happens under the hood"
>         - The pod will be scheduled only on nodes that already contain pods labeled with `app=haproxy`.
> 
>         - If no nodes match, the pod will not be scheduled until a suitable node becomes available.

Reviewer: I don't think this a valid use case :confused: 
I think it's better to go with just 2 use cases that make a lot of sense than trying to squeeze in a 3rd case that doesn't just to show Pod Affinity.
```suggestion
```

=== "Pod anti-affinity"
    ### Pods scheduled apart

    Let's consider a use case in which you want to ensure that no HAProxy pods should be scheduled to run on the same Kubernetes node.

        affinity:
          podAntiAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
            matchExpressions:
                - key: app
                operator: NotIn
                values:
                - haproxy
        topologyKey: "kubernetes.io/hostname"

    ??? info "What happens under the hood"
        - The pod will not be scheduled on nodes that contain pods labeled with `app=haproxy`.

>         - If no nodes match, the pod will not be scheduled until a suitable node becomes available.

Reviewer: Why don't we use the default rule that we set as an example?
Simply setting `topologyKey: "kubernetes.io/hostname"` ensures that the scheduler will try to spread the different DB pods across different k8s nodes, ensuring better fault tolerance in case of node failure.

Author: We can have it explained here itself.

Author: Definitely. It will be more useful.
